### DISCUSSION TITLE
Inferring if a set of lines form a rectangle

### DISCUSSION THREAD

**Discussion by qwertynik** [pymupdf/PyMuPDF#1705] [Looking for help] [answered]
Using the `get_drawings` method fetched all the drawings on a page. Some rectangles, as seen on the PDF, are represented as lines in the output of `get_drawings`. The coordinates can be used to guess if the lines are forming a rectangle. Are there other ways to know with greater accuracy that a group of lines form a rectangle? 

Posting snippet of the output of the `get_drawings` method. In the output 'itemsByShape' and 'items' are additional keys added based on the contents of the `items` key for the path.

```json
{
  "coordinates": {
    "x0": 110.0,
    "y0": 426.3626708984375,
    "x1": 490.0,
    "y1": 474.3626708984375
  },
  "closePath": false,
  "even_odd": false,
  "fill_opacity": 1.0,
  "fill": [
    0.44999998807907104,
    0.08100000023841858,
    0.3109999895095825
  ],
  "seqno": 113,
  "color": null,
  "width": 0,
  "type": "f",
  "lineCap": [
    0
  ],
  "lineJoin": 0,
  "dashes": "[] 0",
  "stroke_opacity": 1,
  "items": [
    {
      "x0": 110.0,
      "y0": 426.3626708984375,
      "x1": 110.0,
      "y1": 426.3626708984375
    },
    {
      "x0": 110.0,
      "y0": 426.3626708984375,
      "x1": 110.0,
      "y1": 474.3626708984375
    },
    {
      "x0": 110.0,
      "y0": 474.3626708984375,
      "x1": 490.0,
      "y1": 474.3626708984375
    },
    {
      "x0": 490.0,
      "y0": 474.3626708984375,
      "x1": 490.0,
      "y1": 426.3626708984375
    }
  ],
  "itemsByShape": {
    "l": [
      0,
      1,
      2,
      3
    ]
  }
}
```

[RectangleDrawnUsingLinesOriginal.pdf](https://github.com/pymupdf/PyMuPDF/files/8649450/RectangleDrawnUsingLinesOriginal.pdf)

Interestingly, the group of lines is recognized as a rectangle in the Foxit PDF editor. Assuming that there should be something pointing towards the fact that it is a rectangle. 
![Demo GIF](https://user-images.githubusercontent.com/6836536/167357179-59cf293a-7121-4048-8b19-eab51af8824e.gif)

**Comment by qwertynik** [comment]
@JorjMcKie 
Would be great if you could add your insights here.

**Reply by qwertynik** [reply]
@JorjMcKie 
Can you add clarity here? That would help. 

Found another use case - triangles are also auto-detected. 
[singleTriangle.pdf](https://github.com/pymupdf/PyMuPDF/files/8725975/singleTriangle.pdf)
![PDFTriangleResize](https://user-images.githubusercontent.com/6836536/169228557-46f2d323-1226-413b-8726-707698b23875.gif)

However, had found only two lines in the path.

```python

import fitz

doc = fitz.open("output/singleTriangle.pdf")
page = doc[0]
paths = page.get_drawings()  # get all drawings on page

print(f"Total paths: {len(paths)}")

# loop through drawings and select relevant ones
for i, path in enumerate(paths):
    pathRect = path["rect"]

    items = path["items"]
    totalItems = len(items)

    print("\n\nPath Details:")
    print(pathRect)
    print('Fill details: ', path['fill'], path['color'])
    print("Total items: ", totalItems)
    for item in path['items']:
        print(item)
```


```
('l', Point(199.07200622558594, 285.02899169921875), Point(210.281005859375, 273.82000732421875))
('l', Point(210.281005859375, 273.82000732421875), Point(199.07200622558594, 262.6109924316406))
```

**Comment by JorjMcKie** [comment, marked as answer]
The underlying data basis of the method only consists of either lines or curves.
Inside the method `get_drawings()`, some logic already identifies rectangles and quads in many, but not all cases.
Detection of other shapes, such as circles, ellipses, triangles, etc. are your burden I am afraid.
I have no insight that goes beyond the trivial / obvious unfortunately.

**Reply by qwertynik** [reply]
The trace output is for the singleTriangle.pdf is below:

```xml
<document filename="/pdfs/singleTriangle.pdf">
<page number="1" mediabox="0 0 600 641">
<set_default_colorspaces gray="DeviceGray" rgb="DeviceRGB" cmyk="DeviceCMYK" oi="None"/>
<fill_path winding="nonzero" colorspace="DeviceRGB" color="0 .2 .627" transform="1 0 0 -1 0 641">
    <moveto x="199.072" y="355.971"/>
    <lineto x="210.281" y="367.18"/>
    <lineto x="199.072" y="378.389"/>
</fill_path>
</page>
</document>
```

Any ideas on how the PDF Editors could be detecting it? Are there any mathematical formulas or rules that can be applied to detect this?

**Reply by qwertynik** [reply]
@JorjMcKie

**Reply by JorjMcKie** [reply]
In the method, the above is translated to three lines with connected end points (plus any stroke / fill colors).
Any drawing / image editing app presumably will contain support for all the basic shapes and be able to track down where lines and curves connect, intersect etc.

**Reply by qwertynik** [reply]
Hmm, makes sense @JorjMcKie
Are you aware of any open-source command-line applications that would be doing this?

**Reply by JorjMcKie** [reply]
No, but it may be worthwhile looking at wxPython. They have a few accompanying tools to make basic drawings.

**Reply by qwertynik** [reply]
Thanks, will checkout. However, it looks like wxPython supports drawing and does not particularly detect the shape a given vector in the PDF.
